src/main/java/picard/sam/markduplicates/UmiGraph.java

```diff
@@ -0,0 +1,210 @@
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package picard.sam.markduplicates;
+
+import htsjdk.samtools.DuplicateSet;
+import htsjdk.samtools.SAMRecord;
+import picard.PicardException;
+
+import java.util.*;
+
+public class UmiGraph {
+    private Map<String, Long> umiCounts;
+    private int[] duplicateSet;    // id[i] = parent of i
+    private Long[] occupancy;
+    private String[] umi;
+    private int editDistanceToJoin;
+    private int nUmis;
+    private int nDuplicateSets;
+
+    UmiGraph() {
+
+    }
+
+    UmiGraph(final Map<String, Long> umiCounts, final int editDistanceToJoin) {
+        // At first every umi is it's own duplicate set
+        nUmis = umiCounts.size();
+        nDuplicateSets = nUmis;
+
+        occupancy = new Long[nUmis];
+        umi = new String[nUmis];
+        this.umiCounts = umiCounts;
+        this.editDistanceToJoin = editDistanceToJoin;
+
+        duplicateSet = new int[nUmis];
+        for (int i = 0; i < nUmis; i++) {
+            duplicateSet[i] = i;
+        }
+
+        int i = 0;
+        for (Map.Entry<String, Long> entry : umiCounts.entrySet()) {
+            umi[i] = entry.getKey();
+            occupancy[i] = entry.getValue();
+            i++;
+        }
+    }
+
+    public int find(int p) {
+        int root = p;
+        while (root != duplicateSet[root])
+            root = duplicateSet[root];
+        while (p != root) {
+            int newp = duplicateSet[p];
+            duplicateSet[p] = root;
+            p = newp;
+        }
+        return root;
+    }
+
+    public void union(int p, int q) {
+        int rootP = find(p);
+        int rootQ = find(q);
+        if (rootP == rootQ) return;
+        duplicateSet[rootP] = rootQ;
+        nDuplicateSets--;
+    }
+
+    List<DuplicateSet> joinUmisIntoDuplicateSets(List<SAMRecord> records, String umiTag, String inferredUmiTag) {
+        // Iterate over diagonal
```

Comment:
Perhaps a better comment is `union all umi pairs that have edit distance less than the max`?